### PR TITLE
[nodefs] Return real values for statvfs via __syscall_statfs64

### PR DESCRIPTION
**Pull request by jeroenpf**
See #22607 

statfs syscall functions are returning hardcoded default values which can in some cases lead to unexpected results.

This PR introduces a way for `__syscall_statfs64` to return real values if the underlying filesystem supports this. For now, I've only implemented this for the NODEFS filesystem but we can expand this to other filesystems when it makes sense.

Additionally, in the previous defaults, `ffree` could be larger than `files` which is incorrect, this has also been addressed.

Ive added a test `test/fs/test_nodefs_statfs.c` which can be executed by running:
`test/runner test_fs_nodefs_statfs`

As I am a new contributor I'm happy to get feedback and make improvements where needed.

### CODE REVIEW

File: src/library_nodefs.js [hunk 1]
```diff
@@ -214,6 +214,13 @@ addToLibrary({
         var path = NODEFS.realPath(node);
         return NODEFS.tryFSOperation(() => fs.readlinkSync(path));
       },
+      statfs(path) {
+        var stats = NODEFS.tryFSOperation(() => fs.statfsSync(path));
```
Comment (sbc100) [inline src/library_nodefs.js:218]: How about this:

```
// Comment about why node does not provide `frsize` but `bsize is a reasonable substitute
stats.frsize = stats.bsize
return stats;
```

File: src/library_fs.js [hunk 2]
```diff
@@ -672,6 +672,31 @@ FS.staticInit();
...
+        flags: 2,
+        namelen: 255,
+      };
+
+      var parent = FS.lookupPath(path, {follow: true}).node;
+      if (typeof parent.node_ops?.statfs !== 'function') {
+        return defaults;
+      }
+
+      return { ...defaults, ...parent.node_ops.statfs(parent.mount.opts.root) };
```
Comment (sbc100) [inline src/library_fs.js:698]: How about:

```
var rtn = {
   ... // defaults
}
var parent = FS.lookupPath(path, {follow: true}).node;
if (typeof parent.node_ops?.statfs) {
   Object.assign(rtn, parent.node_ops.statfs(parent.mount.opts.root));
}
return rtn;
```

Also, so we really need to merge the defaults with the actual results?  Are we expecting backends to only paritally fill the out the information?

File: test/fs/test_nodefs_statfs.c [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+
+int main() {
+    // Test the root filesystem (which should be MEMFS by default)
+    test_statfs("/");
+
+    // Mount NODEFS and test it
+    EM_ASM(
+        FS.mkdir('/working');
+        FS.mount(NODEFS, { root: '.' }, '/working');
+    );
```
Comment (sbc100) [inline test/fs/test_nodefs_statfs.c:30]: Can you move this EM_ASM block into separate function called "setup" and call it at the start on `main`?

File: test/fs/test_nodefs_statfs.c [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+    assert(st.f_blocks > 0 && "Total blocks should be positive");
+    assert(st.f_bfree <= st.f_blocks && "Free blocks should not exceed total blocks");
+    assert(st.f_bavail <= st.f_bfree && "Available blocks should not exceed free blocks");
+    assert(st.f_files > 0 && "Total inodes should be positive");
+    assert(st.f_ffree <= st.f_files && "Free inodes should not exceed total inodes");
+}
+
+int main() {
+    // Test the root filesystem (which should be MEMFS by default)
+    test_statfs("/");
```
Comment (sbc100) [inline test/fs/test_nodefs_statfs.c:24]: Can we call function `test_statvfs` to match the system call name (statvfs)?   Also the filename?

File: test/test_core.py [hunk 5]
```diff
@@ -5779,6 +5779,15 @@ def test_fs_nodefs_readdir(self):
     self.emcc_args += ['-lnodefs.js']
     self.do_runf('fs/test_nodefs_readdir.c', 'success')
 
+  @requires_node
+  def test_fs_nodefs_statfs(self):
+    # externally setup an existing folder structure: existing/a
+    if self.get_setting('WASMFS'):
+      self.set_setting('FORCE_FILESYSTEM')
+    os.makedirs(os.path.join(self.working_dir, 'existing', 'a'))
```
Comment (sbc100) [inline test/test_core.py:5787]: I think the working_dir part here is implicit.. the current working directory will always be the working dir when the test is running.  So I think you can just do `os.makedirs('existing/a')` here.

File: test/test_core.py [hunk 6]
```diff
@@ -5779,6 +5779,15 @@ def test_fs_nodefs_readdir(self):
     self.emcc_args += ['-lnodefs.js']
     self.do_runf('fs/test_nodefs_readdir.c', 'success')
 
+  @requires_node
```
Comment (sbc100) [inline test/test_core.py:5782]: Can you mark this tests as `@crossplatform`?

File: test/test_core.py (hunk 5 above)
Comment (sbc100) [reply test/test_core.py:5787]: (Oh i see this is just copied from above.. I guess that test can be updated too).

File: src/library_fs.js [hunk 7]
```diff
@@ -672,6 +672,31 @@ FS.staticInit();
       }
       return parent.node_ops.mknod(parent, name, mode, dev);
     },
+    statfs(path) {
+
+      // NOTE: None of the defaults here are true. We're just returning safe and
+      //       sane values.
+      var ffree = FS.nextInode - 1; // Free inodes can not be larger than total inodes
```
Comment (sbc100) [inline src/library_fs.js:679]: Maybe just inline this down on line 687?

**Review by sbc100 — COMMENTED** [review]
Thanks for working on this?

I wonder if we can/should make this working under NODERAWFS too?  Or can that be a followup?   I think maybe it would need to happen at the same time?

**Comment by sbc100** [comment]
> Thanks for working on this?
> 
> I wonder if we can/should make this working under NODERAWFS too? Or can that be a followup? I think maybe it would need to happen at the same time?

Actually I think the `NODERAWFS` version can be a followup as it should just keep using the fake/default values up until that point.

File: src/library_fs.js (hunk 2 above)
Comment (jeroenpf) [reply src/library_fs.js]: I've applied the suggested changes.

The idea behind merging the defaults it to ensure that all the required fields are always set. In the case of Node, statfs does not return all of the information:

```
StatFs {
  type: 1397114950,
  bsize: 4096,
  blocks: 121938943,
  bfree: 61058895,
  bavail: 61058895,
  files: 999,
  ffree: 1000000
} 
```

We could add the missing fields in the `node_ops.statfs` but it seemed appropriate to have the FS class ensure a valid response. Im happy to change it though, what do you think?

**Comment by jeroenpf** [comment]
> > Thanks for working on this?
> > I wonder if we can/should make this working under NODERAWFS too? Or can that be a followup? I think maybe it would need to happen at the same time?
> 
> Actually I think the `NODERAWFS` version can be a followup as it should just keep using the fake/default values up until that point.

I agree, all filesystems, besides NODEFS, should continue using the hardcoded values so we can implement the real values for the others in follow-ups.

File: test/fs/test_nodefs_statvfs.c [hunk 8]
```diff
@@ -0,0 +1,35 @@
...
+		FS.mount(NODEFS, { root: '.' }, '/working');
+	);
+}
+
+int main() {
+	setup();
+    // Test the root filesystem (which should be MEMFS by default)
+    test_statvfs("/");
+    test_statvfs("/working");
+    puts("success");
```
Comment (sbc100) [inline test/fs/test_nodefs_statvfs.c:34]: Can you use 2 space indentation in this file?

**Review by sbc100 — APPROVED** [review]

**Comment by sbc100** [comment]
Looks like a test is failing:

```
======================================================================
FAIL [0.000s]: test_codesize_minimal_pthreads (test_other.other)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/usr/lib/python3.8/unittest/case.py", line 60, in testPartExecutor
    yield
  File "/usr/lib/python3.8/unittest/case.py", line 676, in run
    self._callTestMethod(testMethod)
  File "/usr/lib/python3.8/unittest/case.py", line 633, in _callTestMethod
    method()
  File "/root/project/test/common.py", line 330, in decorated
    f(self, *args, **kwargs)
  File "/root/project/test/test_other.py", line 8710, in test_codesize_minimal_pthreads
    self.run_codesize_test('minimal_main.c', ['-Oz', '-pthread', '-sPROXY_TO_PTHREAD'])
  File "/root/project/test/test_other.py", line 8686, in run_codesize_test
    self.assertFileContents(filename, data)
  File "/root/project/test/common.py", line 1480, in assertFileContents
    self.assertTextDataIdentical(expected_content, contents, message,
  File "/root/project/test/common.py", line 1438, in assertTextDataIdentical
    return self.assertIdentical(text1, text2, msg, fromfile, tofile)
  File "/root/project/test/common.py", line 1457, in assertIdentical
    self.fail(fail_message)
  File "/usr/lib/python3.8/unittest/case.py", line 753, in fail
    raise self.failureException(msg)
AssertionError: Unexpected difference:
--- /root/project/test/other/codesize/test_codesize_minimal_pthreads.funcs
+++ /root/project/test/other/codesize/test_codesize_minimal_pthreads.funcs.new
@@ -6,7 +6,6 @@
 $__pthread_mutex_lock
 $__pthread_mutex_trylock
 $__pthread_mutex_unlock
-$__pthread_rwlock_rdlock
 $__pthread_rwlock_timedrdlock
 $__pthread_rwlock_tryrdlock
 $__pthread_rwlock_trywrlock

For full output run with --verbose.
Run with --rebaseline to automatically update expectations

======================================================================
FAIL [0.000s]: test_unistd_fstatfs (test_other.other)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/usr/lib/python3.8/unittest/case.py", line 60, in testPartExecutor
    yield
  File "/usr/lib/python3.8/unittest/case.py", line 676, in run
    self._callTestMethod(testMethod)
  File "/usr/lib/python3.8/unittest/case.py", line 633, in _callTestMethod
    method()
  File "/root/project/test/common.py", line 771, in resulting_test
    return func(self, *args)
  File "/root/project/test/common.py", line 384, in metafunc
    f(self, *args, **kwargs)
  File "/root/project/test/test_other.py", line 13502, in test_unistd_fstatfs
    self.do_run_in_out_file_test('unistd/fstatfs.c')
  File "/root/project/test/common.py", line 1777, in do_run_in_out_file_test
    output = self._build_and_run(srcfile, expected, **kwargs)
  File "/root/project/test/common.py", line 1815, in _build_and_run
    js_output = self.run_js(js_file, engine, args,
  File "/root/project/test/common.py", line 1410, in run_js
    self.fail('JS subprocess failed (%s): %s (expected=%s).  Output:\n%s' % (error.cmd, error.returncode, assert_returncode, ret))
  File "/usr/lib/python3.8/unittest/case.py", line 753, in fail
    raise self.failureException(msg)
AssertionError: JS subprocess failed (/root/emsdk/node/18.20.3_64bit/bin/node --stack-trace-limit=50 --trace-uncaught /tmp/emtest_22wft709/emscripten_test_other_pxlnnu_b/fstatfs.js): 1 (expected=0).  Output:
/tmp/emtest_22wft709/emscripten_test_other_pxlnnu_b/fstatfs.js:2596
        if (parent.node_ops.statfs) {
                   ^

TypeError: Cannot read properties of null (reading 'node_ops')
```

File: test/test_core.py (hunk 6 above)
Comment (Alex089616) [reply test/test_core.py:5782]: test/core/test_statvfs.out

**Review by Alex089616 — CHANGES_REQUESTED** [review]

**Comment by bgrgicak** [comment]
@sbc100 I've taken over this task from @jeroenpf to wrap the work.

I had to [start a new PR to push my fix](https://github.com/emscripten-core/emscripten/pull/22932). When you have time could you please take a look at it?

**Pull request closed without merging**